src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -233,14 +233,9 @@ private class DefaultModelBinderProviderContext : ModelBinderProviderContext
             {
                 _factory = factory;
                 Metadata = factoryContext.Metadata;
-                BindingInfo = new BindingInfo
-                {
-                    BinderModelName = factoryContext.BindingInfo?.BinderModelName ?? Metadata.BinderModelName,
-                    BinderType = factoryContext.BindingInfo?.BinderType ?? Metadata.BinderType,
-                    BindingSource = factoryContext.BindingInfo?.BindingSource ?? Metadata.BindingSource,
-                    PropertyFilterProvider =
-                        factoryContext.BindingInfo?.PropertyFilterProvider ?? Metadata.PropertyFilterProvider,
-                };
+                var bindingInfo = factoryContext.BindingInfo ?? new BindingInfo();
+                bindingInfo.ApplyBindingInfo(Metadata);
+                BindingInfo = bindingInfo;
```

Comment:
The intent is to preserve the pre-2.1 behavior when `AllowValidatingTopLevelNodes` is disabled. We want to continue combining the two even when we could not infer metadata from `BindingInfo` (There's at least one test for this).

Reply:
To clarify my comments:
1. When `AllowValidatingTopLevelNodes` is disabled, the (non-parameter) `ModelMetadata` should already have been used. When does this *not* happen?
2. The pre-2.1 behaviour did _not_ include changing anything in `ModelBinderFactoryContext.BindingInfo` here.

Reply:
So here's the specific test that breaks that's affected: https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.IntegrationTests/BodyValidationIntegrationTests.cs#L833. My reasoning was `ParameterDescriptor.BindingInfo` is settable and you could have `null`-ed it in a convention. For 2.0.0, we would have produced a correct `BindingInfo` inside the `ModelBindingFactory` by applying values from `ModelMetadata`. This was attempting to maintain parity with the behavior.

Reply:
Hmm, actually, this would stomp over known state. I'm going to fix up the test and remove the explicit call here. Seems like an obtuse scenario to support. This should also allow us to make `TryApplyBindingInfo` private

Reply:
Ok, changing that code is going to involve a bunch more test cleanup. We have about a dozen or so Integration tests that start with incomplete or null `BindingInfo` and expect `ModelMetadata` to patch it up.  I'm going to create a follow up to clean this up in rc. For now, I'll leave this code as is just to avoid last minute test clean up churn.

Reply:
Filed https://github.com/aspnet/Mvc/issues/7583